### DISCUSSION TITLE
Raster images set to or near 0 tile level not rendering at a high zoom level 16-20?

### DISCUSSION THREAD

**Discussion by Panthro22** [maplibre/maplibre-gl-js#4108] [General]
When setting an image to a tile level z = 1, why would part of the image applied to the terrain stop rendering when zoomed in to 16-18? I want to understand better some of the constraints applied to raster image source or the constraint affects all layers/sources if set to a low tile level?. The image handles just fine until you zoom in near street level. It disappears for portions of tiles that are too close, even though other tiles further from the camera seem to be rendering fine. Once moved closer to those tiles, it will disappear as well. Is there a cutoff that the child stops rendering an image from a parent tile of it or is there a different constraint?
Zoom out to see image properly render:
![ImageRenderingProperlyAtLowerZoomLevel](https://github.com/maplibre/maplibre-gl-js/assets/41649054/df613de0-41a3-4190-b077-0d92cfd33703)
zoomed in a single tile location is not rendering the image anymore:
![tileSectionNotRenderingTheImage](https://github.com/maplibre/maplibre-gl-js/assets/41649054/d4597d98-a2a5-4ea4-a26b-0db9a99fcb0a)
It seems that I lacking the understanding of why this child tile is not rendering the image anymore?
Is there any constraint of information from a parent tile to how many child tiles the information will be passed down too?
I have noticed if the source is set at tile level 9 then zooming all the way in does not create this problem.

**Comment by HarelM** [comment]
There is a limit to the webgl buffer size I believe and this is causing this issue.
In theory, if the image source would have taken the image and create a tile pyramid from it, it could've solved this issue, but processing may have taken a sometime, depending on the number of zoom levels and tiles that this process needed to generate.
In general, I don't think it makes sense to use image source for an image from the entire world to street view zoom, you either need an image that is huge to begin with, or would have the entire image pixelated in street view zoom values, both options aren't great.
If you need something that wide in terms of range, use a tiled raster source.

**Comment by Panthro22** [comment]
The current problem I am trying to resolve is the images being cut off when the terrain is on. They are bound to a tile that is big enough that the image centered on the tile would fit except, most my images have there centers closer to an edge of that tile and stop rendering properly when you zoom in to that tile level it is assigned. I was trying to resolve this issue, by assigning the image to a tile that the edges of the image would fit into and not be based off of size alone. This works to ensure the image is not to be cut off by a tiles boundary, only downfall now is that the image lands on a low level tile boundary and the image ends up being set to the lowest tile of 0. Which it looks fine and until we zoom in to street level. then depending on its size the whole image could disappear. The images are relatively small, just the bad cases are they have to go to an extreme low tile level to not be chopped off. This solution overall is a short code fix. Unfortunately, zooming in on the small images set to such a low tile level will disappear and stops this solution from being a solid fix. These images are usually assigned to a tile level 10 or greater.

To ensure that I understand what is halting this from being a fix is the image source did not take the image and create a tile pyramid for it or it didn't finish it completely due to image tile level being 0 compared to tile level 16? This range in tile level is creating to large of a buffer for webgl to render the image beyond that point or tile being loaded at 15 or greater? I'm still learning about webgl and mapLibre. Is there a specific spot where I can find the buffer that handles this information? Thank you

**Comment by HarelM** [comment]
There are two problems here: 
1. there's an open bug that image source is being cut when terrain is on
2. You expect too much from the image source - an image source that servers zoom level 0 to 16 is a bad design and you should use a raster tile source instead.

In theory, to solve both problems, it might be a good idea to convert the code under the image source to split the images to tiles and serve them. This is not under development, and only a rough idea on how to solve these issues.

I hope this better explain the situation.
